Modules/arraymodule.c

```diff
@@ -1587,35 +1587,52 @@ static PyObject *
 array_array_tofile_impl(arrayobject *self, PyTypeObject *cls, PyObject *f)
 /*[clinic end generated code: output=4560c628d9c18bc2 input=5a24da7a7b407b52]*/
 {
-    Py_ssize_t nbytes = Py_SIZE(self) * self->ob_descr->itemsize;
     /* Write 64K blocks at a time */
     /* XXX Make the block size settable */
-    int BLOCKSIZE = 64*1024;
+    const Py_ssize_t BLOCKSIZE = 64*1024;
+    const Py_ssize_t itemsize = self->ob_descr->itemsize;
+
+    Py_ssize_t nbytes = Py_SIZE(self) * itemsize;
     Py_ssize_t nblocks = (nbytes + BLOCKSIZE - 1) / BLOCKSIZE;
-    Py_ssize_t i;
 
     if (Py_SIZE(self) == 0)
         goto done;
 
-
     array_state *state = get_array_state_by_class(cls);
     assert(state != NULL);
 
-    for (i = 0; i < nblocks; i++) {
-        char* ptr = self->ob_item + i*BLOCKSIZE;
-        Py_ssize_t size = BLOCKSIZE;
+    for (Py_ssize_t i = 0; i < nblocks; i++) {
+        if (self->ob_item == NULL || Py_SIZE(self) == 0) {
+            break;
+        }
+
+        if (Py_SIZE(self) > PY_SSIZE_T_MAX / itemsize) {
```

Comment:
`Py_SSIZE_T_MAX / itemsize` is a constant that can be computed outside. In addition `Py_SIZE(self)` can be computed once instead of multiple times.

Reply:
Py_SIZE(self) may be changed in iteration, Py_SSIZE_T_MAX / itemsize can set outside the loop